### ISSUE TITLE
Concurrently changing GridifyGlobalConfigurations causes inconsistent behavior

### ISSUE THREAD

**Issue by alirezanet** [alirezanet/Gridify#179]
### Version

v2.14.2

### Details

This issue can be identified and tested within the project's unit tests. Changing the global configuration for a single test can result in other tests' failure, which might also happen in the production environment and actual code base.

### Steps to reproduce

add this test:
e.g:
```csharp
   [Fact]
   public async Task ApplyFiltering_WithoutMapper_ShouldGenerateMappingsForNumbersArray()
   {
      GridifyGlobalConfiguration.DisableNullChecks = true;
      await Task.Delay(20);
      GridifyGlobalConfiguration.DisableNullChecks = false;
   }
```
and run all the unit-tests:

![image](https://github.com/alirezanet/Gridify/assets/7004080/29c4cff2-1b1c-4630-aefa-6c7ac91dd1fb)

**Comment by yasserebrahimi** [comment]
## Additional Analysis & Proposed Solutions

I've done a thorough analysis of this thread safety issue and would like to contribute some detailed solutions.

### Root Cause Analysis
The problem is in `GridifyGlobalConfiguration.cs` where all properties are static without any thread-safety mechanisms:
```csharp
public static bool DisableNullChecks { get; set; } = false;
public static bool CaseInsensitiveFiltering { get; set; } = false;
// ... all other properties
```

### Impact Assessment
- **Severity:** Critical
- **Affected Environments:** All ASP.NET Core applications with concurrent requests
- **Risk:** Race conditions leading to unpredictable query results and potential data corruption

### Proposed Solutions

#### Solution 1: AsyncLocal (Recommended for ASP.NET Core)
```csharp
public static class GridifyGlobalConfiguration
{
    private static AsyncLocal<bool> _disableNullChecks = new(() => false);
    public static bool DisableNullChecks 
    { 
        get => _disableNullChecks.Value;
        set => _disableNullChecks.Value = value;
    }
    
    private static AsyncLocal<bool> _caseInsensitiveFiltering = new(() => false);
    public static bool CaseInsensitiveFiltering 
    { 
        get => _caseInsensitiveFiltering.Value;
        set => _caseInsensitiveFiltering.Value = value;
    }
    
    // Apply to all static properties...
}
```

#### Solution 2: Thread-Safe with Lock (Backward Compatible)
```csharp
public static class GridifyGlobalConfiguration
{
    private static readonly object _lock = new();
    private static bool _disableNullChecks = false;
    
    public static bool DisableNullChecks 
    { 
        get { lock(_lock) return _disableNullChecks; }
        set { lock(_lock) _disableNullChecks = value; }
    }
    // Apply to all properties...
}
```

### Temporary Workaround for Users
Until this is fixed, users should avoid global configuration and use per-query configuration:
```csharp
// Instead of:
GridifyGlobalConfiguration.DisableNullChecks = true; // ❌ Not thread-safe

// Use:
var mapper = new GridifyMapper<T>()
    .Configuration(c => c.DisableNullChecks = true); // ✅ Thread-safe
var result = data.Gridify(query, mapper);
```

I'd be happy to submit a PR if you'd like. Let me know which solution you prefer!

**Comment by alirezanet** [comment]
Hi @ yasserebrahimi 👋,
Thanks for looking into this, For v3, I’m planning to remove the static configuration setup entirely and instead require users to instantiate a new configuration class whenever they want to change defaults.
This way we force explicit ownership of configuration state and avoid thread-safety headaches with shared static state.
Let me know if you see any downside or if there’s a use-case that still requires a static/global config pattern.

**Comment by yasserebrahimi** [comment]
Hi @alirezanet!

**I fully support this approach for v3!** Removing static configuration is the right move for several reasons:

###  Benefits I See:

1. **Thread Safety by Design** - No race conditions possible
2. **Explicit Dependencies** - Makes testing much easier
3. **IoC/DI Friendly** - Works better with dependency injection
4. **Multi-tenant Support** - Different configurations for different tenants
5. **Cleaner Architecture** - Follows SOLID principles

###  Edge Cases to Consider:

#### 1. **Migration Path for v2 → v3**
Many users currently rely on global config. Consider providing:
```csharp
// Backward compatibility helper (with obsolete warning)
[Obsolete("Use GridifyConfiguration instance instead")]
public static class GridifyGlobalConfiguration
{
    public static GridifyConfiguration Default { get; } = new();
    
    // Properties redirect to Default instance
    public static bool DisableNullChecks 
    { 
        get => Default.DisableNullChecks;
        set => Default.DisableNullChecks = value;
    }
}

// Then users can gradually migrate:
// Old way (works but warns):
GridifyGlobalConfiguration.DisableNullChecks = true;

// New way:
var config = new GridifyConfiguration { DisableNullChecks = true };
```

#### 2. **Default Configuration Scenario**
Some users might want a "global default" for their entire app. Maybe provide:
```csharp
public static class Gridify
{
    // Optional: For apps that want one config across the board
    public static GridifyConfiguration DefaultConfiguration { get; set; } 
        = new GridifyConfiguration();
}

// Usage:
// At app startup:
Gridify.DefaultConfiguration = new GridifyConfiguration 
{ 
    DisableNullChecks = true 
};

// Then everywhere:
var result = data.Gridify(query); // Uses Gridify.DefaultConfiguration
// OR
var result = data.Gridify(query, myCustomConfig); // Override
```

#### 3. **ASP.NET Core DI Integration**
For v3, consider adding DI extensions:
```csharp
// Startup.cs / Program.cs
services.AddGridify(config => 
{
    config.DisableNullChecks = true;
    config.CaseInsensitiveFiltering = true;
});

// Then inject:
public class MyController
{
    private readonly IGridifyConfiguration _gridifyConfig;
    
    public MyController(IGridifyConfiguration gridifyConfig)
    {
        _gridifyConfig = gridifyConfig;
    }
}
```

###  Use Cases That *Might* Need Global Config:

1. **Simple Console Apps** - Where DI is overkill
   - **Solution:** Just use a static readonly config instance
   
2. **Legacy Apps** - Already using global config everywhere
   - **Solution:** Migration path with obsolete warnings
   
3. **Prototyping/Testing** - Quick and dirty scripts
   - **Solution:** Provide sensible defaults

###  Recommendation:

For v3:
-  Remove static config (primary API)
-  Make instance-based config the default
-  Provide `GridifyConfiguration.Default` as fallback
-  Add ASP.NET Core DI extensions
-  Write migration guide v2 → v3

For v2 (current):
-  Keep static config but mark as `[Obsolete]` with warning
-  Document the upcoming change
-  Show users the new pattern in docs

### Documentation Needs:

1. **Breaking Changes Guide** - v2 → v3 migration
2. **Best Practices** - How to use instance config
3. **DI Integration** - ASP.NET Core examples
4. **FAQ** - "Why did you remove global config?"

### Bottom Line:

**No major downsides** - this is a solid architectural improvement. The only "downside" is the breaking change, but that's what major versions are for! 

With a good migration path and clear documentation, users will appreciate the cleaner, safer API.

**Happy to help with the v3 implementation or documentation if needed!**
